File: Backend/inventory-service/src/main/java/com/retailnet/inventory/service/SupplierService.java
package com.retailnet.inventory.service;

import com.retailnet.inventory.dto.SupplierDTO;
import java.util.List;

/**
 * Interface defining the business contract for Supplier management.
 * Uses DTOs to ensure the API layer never touches raw Database Entities.
 */
public interface SupplierService {

    /**
     * Registers a new vendor in the system.
     * @param supplierDTO Data from the API request
     * @return The saved Supplier data including its generated ID
     */
    SupplierDTO addSupplier(SupplierDTO supplierDTO);

    /**
     * Retrieves all registered suppliers for the UI.
     * @return List of SupplierDTOs
     */
    List<SupplierDTO> findAllSuppliers();

    /**
     * Updates an existing supplier's details.
     *
     * @param id The ID of the supplier to update
     * @param supplierDTO The updated supplier data
     * @return The updated SupplierDTO
     */
    SupplierDTO updateSupplier(Long id, SupplierDTO supplierDTO);

    /**
     * Deletes a supplier from the database.
     *
     * @param id The ID of the supplier to delete
     */
    void deleteSupplier(Long id);
}

File: ui/retailnet-workspace/retailnet-ui/src/app/shared/services/supplier.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { AppContextService } from './app-context.service';
import { SupplierDTO } from '../models/inventory.models';

@Injectable({
  providedIn: 'root'
})
export class SupplierService {
  private readonly http = inject(HttpClient);
  private readonly appContext = inject(AppContextService);

  private readonly httpOptions = {
    headers: new HttpHeaders({ 'Content-Type': 'application/json' })
  };

  private getUrl(endpoint: string): string {
    return this.appContext.getContext('INVENTORY_CONTEXT') + endpoint;
  }

  /**
   * Retrieves a comprehensive list of all currently registered suppliers.
   * Used for population of vendor selection lists and reporting.
   * Maps to GET /api/suppliers/all
   */
  getSuppliers(): Observable<SupplierDTO[]> {
    return this.http.get<SupplierDTO[]>(this.getUrl('suppliers/all'));
  }

  /**
   * Registers a new supplier in the system.
   * Consumes JSON data to create a persistent supplier record.
   * Maps to POST /api/suppliers/add
   */
  addSupplier(supplier: SupplierDTO): Observable<SupplierDTO> {
    return this.http.post<SupplierDTO>(this.getUrl('suppliers/add'), supplier, this.httpOptions);
  }

  /**
   * Updates an existing supplier's information.
   * Maps to PUT /api/suppliers/update/{id}
   */
  updateSupplier(id: number, supplier: SupplierDTO): Observable<SupplierDTO> {
    return this.http.put<SupplierDTO>(this.getUrl(`suppliers/update/${id}`), supplier, this.httpOptions);
  }

  /**
   * Removes a supplier from the system.
   * Maps to DELETE /api/suppliers/delete/{id}
   */
  deleteSupplier(id: number): Observable<void> {
    return this.http.delete<void>(this.getUrl(`suppliers/delete/${id}`));
  }
}
